util.c

```diff
@@ -783,7 +783,66 @@ size_t filter_result_len(const char *filter, const char *user, char *output) {
 }
 
 char *filter_printf(const char *filter, const char *user) {
-  char *result = malloc(filter_result_len(filter, user, NULL));
-  filter_result_len(filter, user, result);
+  char *user_rfc4515 = rfc4515_replace(user);
+  char *result = malloc(filter_result_len(filter, user_rfc4515, NULL));
+  filter_result_len(filter, user_rfc4515, result);
+  free(user_rfc4515);
   return result;
 }
+
+size_t rfc4515_length (const char* in, char* out)
+{
+    const char* pos = NULL;
+    size_t result = 0;
+    const char specials[] = { 0x5c, 0x2a, 0x28, 0x29, '\n' };
```

Comment:
should \n be escaped here, also the string still doesn't have a NULL at the end.
Don't you think something like `const char specials[] = "\*()";` is clearer?

Also.. Reading rfc4515 are you sure that the escaping of the string "()" should be "\(\)" shouldn't it be "\28\29" ?

Reply:
Hey @klali, 

> should \n be escaped here, also the string still doesn't have a NULL at the end.
damn it! `\n` is wrong. `\0` is right. My mistake. Sorry, it was just too late last night.

> Don't you think something like `const char specials[] = "\\*()";` is clearer?
I chose the hexadecimal notation because that's how it is specified in the RFC. But of course a backslash corresponds to 0x5c, so I can change that too. 

> Also.. Reading rfc4515 are you sure that the escaping of the string "()" should be "\(\)" shouldn't it be "\28\29" ?
I think it doesn't make sense if the LDAP filter contains "EXAMPLE\5cjohn" (5c corresponds to the hexadecimal value of the backslash) instead of "EXAMPLE\john". As I understand it, the characters must be preceded by a backslash. Similar to C ++.

I compiled the PAM module with my changes last night and installed it on my Samba AD DC. It worked.

Reply:
Hey @klali,

I checked RFC4515 again. According to [RFC4515, Chapter 4](https://tools.ietf.org/html/rfc4515#section-4) you are right. `\\` must be replaced with `\\5C`. The only thing that irritates me is that it also works with `\\\\`. At least this was shown by the test on my Samba AD DC yesterday evening.

Reply:
Two discussions in one, sorry about that, I should've made two comments.

Regarding hex/values in the string I don't really care, as you've noticed it's fragile to specify them as array members, it's easy to forget the NULL termination and strpbrk() will eat a string. another option is "\x5c\x2a\x28\x29". If you have them as hex it needs a comment what that corresponds to.

Regarding how it should be correctly represented I think the rfc is clear that it should be "\ hex hex", but more research might be needed to understand what's most portable.

Reply:
Hey @klali, I think it makes sense to comply with the RFC4515. For this reason I have adapted my code and replaced the characters according to the specifications. In addition, I recompiled the PAM module and tested it on my SAMBA AD DC. It works too. I would prefer to use this new version because it implements the specifications of RFC4515.